### PR TITLE
feat: add "add binding" ui to vscode extension

### PR DESCRIPTION
**Pull request by emily-shen**
Fixes [DEVX-1473](https://jira.cfdata.org/browse/DEVX-1473)

---

- Tests
  - [ ] TODO (before merge)
  - [ ] Tests included
  - [x] Tests not necessary because: will do in a followup (haven't set up CI tests for this yet anyway), has been manually tested
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [ ] Required
  - [x] Not required because: no relevant e2e's
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): 
  - [x] Documentation not necessary because: not released yet

### CODE REVIEW

**Comment by CarmenPopoviciu** [comment]
@emily-shen why are we doing the tests in a follow-up?

File: packages/cloudflare-workers-bindings-extension/src/wrangler.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import * as path from "path";
+import * as vscode from "vscode";
```
Comment (penalosa) [inline packages/cloudflare-workers-bindings-extension/src/wrangler.ts:2]: What's this for?

File: packages/cloudflare-workers-bindings-extension/src/extension.ts [hunk 2]
```diff
@@ -32,6 +33,13 @@ export async function activate(
 		() => bindingsProvider.refresh()
 	);
 
+	// Register the add bindings command
+	vscode.commands.registerCommand(
```
Comment (penalosa) [inline packages/cloudflare-workers-bindings-extension/src/extension.ts:37]: We should add this to the context subscription list below so it's cleaned up properly

File: packages/cloudflare-workers-bindings-extension/src/add-binding.ts [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+class BindingType implements QuickPickItem {
+	constructor(
+		public label: string,
+		public description?: string,
+		public detail?: string,
+		public iconPath?: Uri
+	) {}
+}
+
+export async function multiStepInput(context: ExtensionContext) {
```
Comment (penalosa) [inline packages/cloudflare-workers-bindings-extension/src/add-binding.ts:25]: Could you give this a more descriptive name?

**Comment by emily-shen** [comment]
> @emily-shen why are tests being handled in a follow-up?

we want to get this out before the LDW, and the extension is not yet released, and the vscode extension testing process if a bit undefined for us still and we can't run them in CI yet :')

File: packages/cloudflare-workers-bindings-extension/src/add-binding.ts [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+	window,
+	workspace,
+} from "vscode";
+import { getConfigUri } from "./bindings";
+import { importWrangler } from "./wrangler";
+
+class BindingType implements QuickPickItem {
+	constructor(
+		public label: string,
+		public description?: string,
```
Comment (penalosa) [inline packages/cloudflare-workers-bindings-extension/src/add-binding.ts:19]: I'm not sure `description` is right here—maybe `configKey`?

File: packages/cloudflare-workers-bindings-extension/src/add-binding.ts [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+			value: state.name || "",
+			prompt: "Choose a binding name",
+			validate: validateNameIsUnique,
+			placeholder: `e.g. MY_BINDING`,
+		});
+		state.name = name;
+		return () => addToToml(state);
+	}
+
+	async function addToToml(state: Partial<State>) {
```
Comment (penalosa) [inline packages/cloudflare-workers-bindings-extension/src/add-binding.ts:96]:
```suggestion
	async function addToConfig(state: Partial<State>) {
```

File: packages/cloudflare-workers-bindings-extension/src/add-binding.ts [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+		const configUri = await getConfigUri();
+		if (!configUri) {
+			return null;
```
Comment (penalosa) [inline packages/cloudflare-workers-bindings-extension/src/add-binding.ts:99]: This should probably show an error notification along the lines of "Unable to locate Wrangler configuration file—have you opened a Wrangler project?"

File: packages/cloudflare-workers-bindings-extension/src/add-binding.ts [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+		}
+		const workspaceFolder = workspace.getWorkspaceFolder(configUri);
+
+		if (!workspaceFolder) {
+			return null;
+		}
+
+		const wrangler = importWrangler(workspaceFolder.uri.fsPath);
+
+		await workspace.openTextDocument(configUri).then((doc) => {
```
Comment (penalosa) [inline packages/cloudflare-workers-bindings-extension/src/add-binding.ts:109]: Why is this in a `.then()` if `openTextDocument` is awaited??

File: packages/cloudflare-workers-bindings-extension/src/add-binding.ts [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+binding = "${state.name}"
+`;
+
+				env.clipboard.writeText(patch);
+			}
+		});
+	}
+
+	async function validateNameIsUnique(name: string) {
+		// TODO: actually validate uniqueness
```
Comment (penalosa) [inline packages/cloudflare-workers-bindings-extension/src/add-binding.ts:131]: Could you make an issue to track this?

File: packages/cloudflare-workers-bindings-extension/src/wrangler.ts (hunk 1 above)
Comment (emily-shen) [reply packages/cloudflare-workers-bindings-extension/src/wrangler.ts]: [cc97bed](https://github.com/cloudflare/workers-sdk/pull/7591/commits/cc97bed3cb7af655d8f05c2e2441b60cda85c1c3)

File: packages/cloudflare-workers-bindings-extension/src/add-binding.ts (hunk 3 above)
Comment (emily-shen) [reply packages/cloudflare-workers-bindings-extension/src/add-binding.ts]: [cc97bed](https://github.com/cloudflare/workers-sdk/pull/7591/commits/cc97bed3cb7af655d8f05c2e2441b60cda85c1c3)

File: packages/cloudflare-workers-bindings-extension/src/add-binding.ts (hunk 7 above)
Comment (emily-shen) [reply packages/cloudflare-workers-bindings-extension/src/add-binding.ts]: [cc97bed](https://github.com/cloudflare/workers-sdk/pull/7591/commits/cc97bed3cb7af655d8f05c2e2441b60cda85c1c3)

File: packages/cloudflare-workers-bindings-extension/src/add-binding.ts (hunk 6 above)
Comment (emily-shen) [reply packages/cloudflare-workers-bindings-extension/src/add-binding.ts:104]: [cc97bed](https://github.com/cloudflare/workers-sdk/pull/7591/commits/cc97bed3cb7af655d8f05c2e2441b60cda85c1c3)

File: packages/cloudflare-workers-bindings-extension/src/add-binding.ts (hunk 8 above)
Comment (emily-shen) [reply packages/cloudflare-workers-bindings-extension/src/add-binding.ts:136]: [DEVX-1547](https://jira.cfdata.org/browse/DEVX-1547)

**Review by penalosa — APPROVED** [review]

**Pull request merged**